File: TorrentAdder/SYComputerFormCell.m
//
//  SYComputerFormCell.m
//  TorrentAdder
//
//  Created by Stan Chevallier on 05/10/2015.
//  Copyright © 2015 Syan. All rights reserved.
//

#import "SYComputerFormCell.h"
#import "SYComputerModel.h"
#import "UIImage+SY.h"
#import "UIColor+SY.h"

@interface SYComputerFormCell () <UITextFieldDelegate>
@property (nonatomic, weak) IBOutlet UIImageView *iconView;
@property (nonatomic, weak) IBOutlet UITextField *textField;
@property (nonatomic, weak) IBOutlet UISegmentedControl *segmentedControl;
@property (nonatomic, assign) SYComputerModelField  field;
@property (nonatomic, strong) SYComputerModel       *computer;
@end

@implementation SYComputerFormCell

- (void)setComputer:(SYComputerModel *)computer andField:(SYComputerModelField)field
{
    [self setComputer:computer];
    [self setField:field];
    
    [self.iconView setImage:[[computer imageForField:field] imageMaskedWithColor:[UIColor lightBlueColor]]];
    [self.textField setKeyboardType:[computer keyboardTypeForField:field]];
    
    NSArray *options = [computer optionsForEnumField:field];
    if (options)
    {
        [self.segmentedControl setHidden:NO];
        [self.segmentedControl removeAllSegments];
        for (NSUInteger i = 0; i < [options count]; ++i)
            [self.segmentedControl insertSegmentWithTitle:options[i] atIndex:i animated:NO];
        [self.segmentedControl setSelectedSegmentIndex:[[computer valueForField:field] intValue]];
        [self.textField setHidden:YES];
    }
    else
    {
        [self.segmentedControl setHidden:YES];
        [self.textField setHidden:NO];
        [self.textField setPlaceholder:[computer nameForField:field]];
        [self.textField setText:[computer valueForField:field]];
        [self.textField setTextAlignment:NSTextAlignmentRight];
    }
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [self.computer setValue:textField.text forField:self.field];
    [textField resignFirstResponder];
    return NO;
}

- (IBAction)textFieldTextUpdated:(UITextField *)textField
{
    [self.computer setValue:textField.text forField:self.field];
}

- (IBAction)segmentedControlUpdated:(UISegmentedControl *)control
{
    [self.computer setValue:@(control.selectedSegmentIndex) forField:self.field];
}

@end

File: TorrentAdder/SYComputerFormCell.swift
//
//  SYComputerFormCell.swift
//  TorrentAdder
//
//  Created by Stanislas Chevallier on 28/11/2018.
//  Copyright © 2018 Syan. All rights reserved.
//

import UIKit

class SYComputerFormCell: UITableViewCell {
    
    // MARK: Init
    override func awakeFromNib() {
        super.awakeFromNib()
        segmentedControl.tintColor = .lightBlue
    }

    // MARK: Properties
    var formField: SYClient.FormField = .host {
        didSet {
            updateContent()
        }
    }
    var computer: SYClient? {
        didSet {
            updateContent()
        }
    }
    
    // MARK: Views
    @IBOutlet private var iconView: UIImageView!
    @IBOutlet private var textField: UITextField!
    @IBOutlet private var segmentedControl: UISegmentedControl!

    // MARK: Actions
    @IBAction private func segmentedControlChanged() {
        computer?.setValue(segmentedControl.selectedSegmentIndex, for: formField)
    }
    
    @IBAction private func textFieldEditingChanged() {
        computer?.setValue(textField.text ?? "", for: formField)
    }
    
    // MARK: Content
    private func updateContent() {
        guard let computer = computer else { return }
        
        iconView.image = formField.image?.sy_imageMasked(with: .lightBlue)
        textField.keyboardType = formField.keyboardType
        if #available(iOS 11.0, *) {
            textField.textContentType = formField.textContentType
        }

        if let options = formField.options, !options.isEmpty {
            textField.isHidden = true
            segmentedControl.isHidden = false
            segmentedControl.removeAllSegments()
            options.keys.sorted().forEach { (index) in
                segmentedControl.insertSegment(withTitle: options[index], at: index, animated: false)
            }
            segmentedControl.selectedSegmentIndex = computer.intValue(for: formField) ?? 0
        }
        else {
            textField.isHidden = false
            segmentedControl.isHidden = true
            textField.placeholder = formField.name
            textField.text = computer.stringValue(for: formField)
        }
    }
}

extension SYComputerFormCell : UITextFieldDelegate {
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        computer?.setValue(textField.text ?? "", for: formField)
        textField.resignFirstResponder()
        return false
    }
}
